File: src/modules/review/review.service.ts
import { Types } from 'mongoose';
import { AppError } from '../../shared/errors/AppError.js';
import { recordAudit } from '../../shared/utils/audit.js';
import {
  buildPaginationMeta,
  toPageOptions,
  type PaginationMeta,
} from '../../shared/utils/pagination.js';
import { Product } from '../catalog/product.model.js';
import { Order } from '../order/order.model.js';
import { User } from '../user/user.model.js';
import { Review, type ReviewStatus } from './review.model.js';

export interface ReviewPublic {
  id: string;
  productId: string;
  userId: string;
  reviewerName: string;
  rating: number;
  title: string;
  body: string;
  isVerifiedPurchase: boolean;
  status: ReviewStatus;
  moderationReason: string | null;
  createdAt: string;
  updatedAt: string;
}

interface ReviewRow {
  _id: unknown;
  productId: { toString(): string };
  userId: { toString(): string };
  rating: number;
  title: string;
  body: string;
  isVerifiedPurchase: boolean;
  status: ReviewStatus;
  moderationReason: string | null;
  createdAt: Date;
  updatedAt: Date;
  reviewer?: { name?: string | null } | null;
}

function toReviewPublic(r: ReviewRow): ReviewPublic {
  return {
    id: (r._id as { toString(): string }).toString(),
    productId: r.productId.toString(),
    userId: r.userId.toString(),
    reviewerName: r.reviewer?.name ?? 'Anonymous',
    rating: r.rating,
    title: r.title,
    body: r.body,
    isVerifiedPurchase: r.isVerifiedPurchase,
    status: r.status,
    moderationReason: r.moderationReason,
    createdAt: r.createdAt.toISOString(),
    updatedAt: r.updatedAt.toISOString(),
  };
}

// ---------------------------------------------------------------------------
// Create
// ---------------------------------------------------------------------------

export async function createReview(
  userId: string,
  productId: string,
  input: { rating: number; title?: string; body: string },
): Promise<ReviewPublic> {
  const product = await Product.exists({ _id: productId, status: 'published', isActive: true });
  if (!product) throw AppError.notFound('Product not found');

  const existing = await Review.exists({ productId, userId });
  if (existing) {
    throw AppError.conflict('You have already reviewed this product', {
      errorCode: 'REVIEW_EXISTS',
    });
  }

  // Verified purchases are auto-approved; unreviewed-by-purchase submissions
  // are moderated. Look for a delivered order containing this product.
  const verifiedOrder = await Order.findOne({
    userId,
    status: 'delivered',
    'items.productId': productId,
  }).lean();

  const review = await Review.create({
    productId,
    userId,
    orderId: verifiedOrder ? verifiedOrder._id : null,
    rating: input.rating,
    title: input.title ?? '',
    body: input.body,
    isVerifiedPurchase: verifiedOrder !== null,
    status: verifiedOrder ? 'approved' : 'pending',
  });

  recordAudit(
    'review.created',
    { reviewId: review._id.toString(), productId },
    { actorId: userId },
  );
  return toReviewPublic(review.toObject() as unknown as ReviewRow);
}

export async function updateOwnReview(
  userId: string,
  reviewId: string,
  input: { rating?: number; title?: string; body?: string },
): Promise<ReviewPublic> {
  const review = await Review.findOne({ _id: reviewId, userId });
  if (!review) throw AppError.notFound('Review not found');

  if (input.rating !== undefined) review.rating = input.rating;
  if (input.title !== undefined) review.title = input.title;
  if (input.body !== undefined) review.body = input.body;
  await review.save();

  recordAudit('review.updated', { reviewId }, { actorId: userId });
  return toReviewPublic(review.toObject() as unknown as ReviewRow);
}

export async function deleteReview(
  userId: string,
  reviewId: string,
  isStaff: boolean,
): Promise<void> {
  const review = isStaff
    ? await Review.findById(reviewId)
    : await Review.findOne({ _id: reviewId, userId });
  if (!review) throw AppError.notFound('Review not found');
  await review.deleteOne();
  recordAudit('review.deleted', { reviewId }, { actorId: userId });
}

// ---------------------------------------------------------------------------
// Reads
// ---------------------------------------------------------------------------

export async function listReviews(
  productId: string,
  query: { page?: number; limit?: number; status?: ReviewStatus },
  isStaff: boolean,
): Promise<{ reviews: ReviewPublic[]; meta: PaginationMeta }> {
  const filter: Record<string, unknown> = { productId };
  // Public listings only show approved reviews; staff may filter by status.
  if (isStaff) {
    if (query.status) filter.status = query.status;
  } else {
    filter.status = 'approved';
  }

  const options = toPageOptions(query);
  const [total, reviews] = await Promise.all([
    Review.countDocuments(filter),
    Review.find(filter).sort({ createdAt: -1 }).skip(options.skip).limit(options.limit).lean(),
  ]);

  const userIds = [...new Set(reviews.map((r) => r.userId.toString()))];
  const users = await User.find({ _id: { $in: userIds } })
    .select('name')
    .lean();
  const nameById = new Map(users.map((u) => [u._id.toString(), u.name]));

  return {
    reviews: reviews.map((r) =>
      toReviewPublic({ ...r, reviewer: { name: nameById.get(r.userId.toString()) ?? null } }),
    ),
    meta: buildPaginationMeta(total, options),
  };
}

export async function productRatingSummary(productId: string): Promise<{
  average: number;
  count: number;
  distribution: Record<number, number>;
}> {
  const rows = await Review.aggregate([
    { $match: { productId: new Types.ObjectId(productId), status: 'approved' } },
    {
      $group: {
        _id: '$rating',
        count: { $sum: 1 },
      },
    },
  ]);

  const distribution: Record<number, number> = { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 };
  let total = 0;
  let weighted = 0;
  for (const row of rows as Array<{ _id: number; count: number }>) {
    distribution[row._id] = row.count;
    total += row.count;
    weighted += row._id * row.count;
  }
  return { average: total > 0 ? weighted / total : 0, count: total, distribution };
}

// ---------------------------------------------------------------------------
// Moderation (staff)
// ---------------------------------------------------------------------------

export async function moderateReview(
  reviewId: string,
  action: 'approve' | 'reject',
  reason: string,
  actorId: string,
): Promise<ReviewPublic> {
  const review = await Review.findById(reviewId);
  if (!review) throw AppError.notFound('Review not found');

  review.status = action === 'approve' ? 'approved' : 'rejected';
  review.moderationReason = action === 'reject' ? reason || 'Rejected by moderator' : null;
  await review.save();

  recordAudit('review.moderated', { reviewId, action, reason }, { actorId });
  return toReviewPublic(review.toObject() as unknown as ReviewRow);
}

// Internal: checked by admin dashboards (top rated, lowest rated).
export async function listAllReviews(query: {
  page?: number;
  limit?: number;
  status?: ReviewStatus;
}): Promise<{ reviews: ReviewPublic[]; meta: PaginationMeta }> {
  const filter: Record<string, unknown> = {};
  if (query.status) filter.status = query.status;
  const options = toPageOptions(query);
  const [total, reviews] = await Promise.all([
    Review.countDocuments(filter),
    Review.find(filter).sort({ createdAt: -1 }).skip(options.skip).limit(options.limit).lean(),
  ]);
  const userIds = [...new Set(reviews.map((r) => r.userId.toString()))];
  const users = await User.find({ _id: { $in: userIds } })
    .select('name')
    .lean();
  const nameById = new Map(users.map((u) => [u._id.toString(), u.name]));
  return {
    reviews: reviews.map((r) =>
      toReviewPublic({ ...r, reviewer: { name: nameById.get(r.userId.toString()) ?? null } }),
    ),
    meta: buildPaginationMeta(total, options),
  };
}

File: java/src/main/java/com/ecommerce/backend/modules/review/ReviewService.java
package com.ecommerce.backend.modules.review;

import com.ecommerce.backend.common.error.ApiException;
import com.ecommerce.backend.modules.audit.AuditService;
import com.ecommerce.backend.modules.catalog.Product;
import com.ecommerce.backend.modules.order.Order;
import com.ecommerce.backend.modules.user.User;
import com.ecommerce.backend.modules.user.UserRepository;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.bson.Document;
import org.bson.types.ObjectId;
import org.springframework.data.domain.Sort;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.aggregation.Aggregation;
import org.springframework.data.mongodb.core.aggregation.AggregationResults;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.stereotype.Service;

/** Review lifecycle (mirrors {@code review.service.ts}). */
@Service
public class ReviewService {

    private final ReviewRepository reviewRepository;
    private final UserRepository userRepository;
    private final MongoTemplate mongo;
    private final AuditService auditService;

    public ReviewService(
            ReviewRepository reviewRepository,
            UserRepository userRepository,
            MongoTemplate mongo,
            AuditService auditService) {
        this.reviewRepository = reviewRepository;
        this.userRepository = userRepository;
        this.mongo = mongo;
        this.auditService = auditService;
    }

    // -------------------------------------------------------------- Public shapes

    public record ReviewPublic(
            String id,
            String productId,
            String userId,
            String reviewerName,
            int rating,
            String title,
            String body,
            boolean isVerifiedPurchase,
            String status,
            String moderationReason,
            String createdAt,
            String updatedAt) {}

    public record ReviewListResult(List<ReviewPublic> reviews, long total) {}

    public record RatingSummary(double average, long count, Map<Integer, Integer> distribution) {}

    // ----------------------------------------------------------------- Create

    public ReviewPublic createReview(String userId, String productId, ReviewDtos.CreateReviewRequest input) {
        Product product = mongo.findOne(
                Query.query(Criteria.where("_id")
                        .is(productId)
                        .and("status")
                        .is(Product.STATUS_PUBLISHED)
                        .and("isActive")
                        .is(true)),
                Product.class);
        if (product == null) {
            throw ApiException.notFound("Product not found");
        }

        if (reviewRepository.existsByProductIdAndUserId(productId, userId)) {
            throw ApiException.conflict(
                    "You have already reviewed this product", "REVIEW_EXISTS");
        }

        Order verifiedOrder = mongo.findOne(
                Query.query(Criteria.where("userId")
                        .is(userId)
                        .and("status")
                        .is(Order.STATUS_DELIVERED)
                        .and("items.productId")
                        .is(productId)),
                Order.class);

        Review review = new Review();
        review.productId = productId;
        review.userId = userId;
        review.orderId = verifiedOrder == null ? null : verifiedOrder.id;
        review.rating = input.rating();
        review.title = input.title() == null ? "" : input.title().trim();
        review.body = input.body().trim();
        review.isVerifiedPurchase = verifiedOrder != null;
        review.status = verifiedOrder == null ? Review.STATUS_PENDING : Review.STATUS_APPROVED;
        review.createdAt = Instant.now();
        review.updatedAt = review.createdAt;
        Review saved = reviewRepository.insert(review);

        auditService.log("review.created", userId, null,
                Map.of("reviewId", saved.id, "productId", productId));
        return toReviewPublic(saved, null);
    }

    // ------------------------------------------------------------- Own review

    public ReviewPublic updateOwnReview(String userId, String reviewId, ReviewDtos.UpdateReviewRequest input) {
        if (input.rating() == null && input.title() == null && input.body() == null) {
            throw ApiException.badRequest("At least one field is required", "BAD_REQUEST");
        }
        Review review = findOneByUser(reviewId, userId);
        if (input.rating() != null) {
            review.rating = input.rating();
        }
        if (input.title() != null) {
            review.title = input.title().trim();
        }
        if (input.body() != null) {
            review.body = input.body().trim();
        }
        review.updatedAt = Instant.now();
        Review saved = reviewRepository.save(review);
        auditService.log("review.updated", userId, null, Map.of("reviewId", reviewId));
        return toReviewPublic(saved, reviewerName(userId));
    }

    public void deleteReview(String userId, String reviewId, boolean isStaff) {
        Review review = isStaff
                ? findReview(reviewId)
                : findOneByUser(reviewId, userId);
        reviewRepository.delete(review);
        auditService.log("review.deleted", userId, null, Map.of("reviewId", reviewId));
    }

    // ----------------------------------------------------------------- Reads

    public ReviewListResult listReviews(String productId, int page, int limit, String status, boolean isStaff) {
        Query query = new Query(Criteria.where("productId").is(productId));
        if (isStaff) {
            if (status != null && !status.isBlank()) {
                query.addCriteria(Criteria.where("status").is(status));
            }
        } else {
            query.addCriteria(Criteria.where("status").is(Review.STATUS_APPROVED));
        }
        return listWithNames(query, page, limit);
    }

    public ReviewListResult listAllReviews(int page, int limit, String status) {
        Query query = new Query();
        if (status != null && !status.isBlank()) {
            query.addCriteria(Criteria.where("status").is(status));
        }
        return listWithNames(query, page, limit);
    }

    private ReviewListResult listWithNames(Query query, int page, int limit) {
        long total = mongo.count(query, Review.class);
        query.with(Sort.by(Sort.Direction.DESC, "createdAt"))
                .skip((long) (page - 1) * limit)
                .limit(limit);
        List<Review> reviews = mongo.find(query, Review.class);
        Map<String, String> nameById = namesFor(reviews.stream().map(r -> r.userId).toList());
        return new ReviewListResult(
                reviews.stream().map(r -> toReviewPublic(r, nameById.get(r.userId))).toList(),
                total);
    }

    public RatingSummary productRatingSummary(String productId) {
        Aggregation aggregation = Aggregation.newAggregation(
                Aggregation.match(Criteria.where("productId")
                        .is(productId)
                        .and("status")
                        .is(Review.STATUS_APPROVED)),
                Aggregation.group("rating").count().as("count"));
        AggregationResults<Document> results = mongo.aggregate(aggregation, Review.class, Document.class);

        Map<Integer, Integer> distribution = new LinkedHashMap<>();
        for (int i = 1; i <= Review.MAX_RATING; i++) {
            distribution.put(i, 0);
        }
        long count = 0;
        long weighted = 0;
        for (Document row : results.getMappedResults()) {
            int rating = ((Number) row.get("_id")).intValue();
            int rowCount = ((Number) row.get("count")).intValue();
            distribution.put(rating, rowCount);
            count += rowCount;
            weighted += (long) rating * rowCount;
        }
        double average = count > 0 ? (double) weighted / count : 0;
        return new RatingSummary(average, count, distribution);
    }

    // ------------------------------------------------------------ Moderation

    public ReviewPublic moderateReview(String reviewId, String action, String reason, String actorId) {
        Review review = findReview(reviewId);
        review.status = "approve".equals(action) ? Review.STATUS_APPROVED : Review.STATUS_REJECTED;
        review.moderationReason = "reject".equals(action)
                ? (reason == null || reason.isBlank() ? "Rejected by moderator" : reason)
                : null;
        review.updatedAt = Instant.now();
        Review saved = reviewRepository.save(review);
        auditService.log("review.moderated", actorId, null,
                Map.of("reviewId", reviewId, "action", action, "reason", reason == null ? "" : reason));
        return toReviewPublic(saved, reviewerName(saved.userId));
    }

    // -------------------------------------------------------------- Internals

    private Review findReview(String reviewId) {
        if (!ObjectId.isValid(reviewId)) {
            throw ApiException.badRequest("Invalid identifier format", "BAD_REQUEST");
        }
        return reviewRepository.findById(reviewId)
                .orElseThrow(() -> ApiException.notFound("Review not found"));
    }

    private Review findOneByUser(String reviewId, String userId) {
        Review review = findReview(reviewId);
        if (!userId.equals(review.userId)) {
            throw ApiException.notFound("Review not found");
        }
        return review;
    }

    private String reviewerName(String userId) {
        if (userId == null) {
            return "Anonymous";
        }
        return userRepository.findById(userId).map(u -> u.name).orElse("Anonymous");
    }

    private Map<String, String> namesFor(List<String> userIds) {
        List<String> distinct = userIds.stream().distinct().toList();
        if (distinct.isEmpty()) {
            return Map.of();
        }
        Map<String, String> names = new java.util.HashMap<>();
        for (User u : userRepository.findAllById(distinct)) {
            names.put(u.id, u.name == null ? "" : u.name);
        }
        return names;
    }

    private ReviewPublic toReviewPublic(Review r, String reviewerName) {
        return new ReviewPublic(
                r.id,
                r.productId,
                r.userId,
                reviewerName == null || reviewerName.isBlank() ? "Anonymous" : reviewerName,
                r.rating,
                r.title == null ? "" : r.title,
                r.body,
                r.isVerifiedPurchase,
                r.status,
                r.moderationReason,
                iso(r.createdAt),
                iso(r.updatedAt));
    }

    private static String iso(Instant value) {
        return value == null ? null : value.toString();
    }
}
